### ISSUE TITLE
test_underUnderPathAlreadyImported and test_listingModules from twisted.test.test_modules.(Zip)PathModificationTest fail with Gtk2 reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2139]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2139|
|Type|enhancement|
|Created|2006-09-30 22:15:07Z|

```
===============================================================================
[ERROR]: twisted.test.test_modules.PathModificationTest.test_listingModules

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 163, in test_listingModules
    self._listModules()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 152, in _listModules
    pkginfo = modules.getModule(self.packageName)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 587, in __getitem__
    raise KeyError(modname)
exceptions.KeyError: 'pyspacetests0'
===============================================================================
[ERROR]: twisted.test.test_modules.PathModificationTest.test_underUnderPathAlreadyImported

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 138, in test_underUnderPathAlreadyImported
    self._underUnderPathTest()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 127, in _underUnderPathTest
    modinfo = modules.getModule(self.packageName)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 587, in __getitem__
    raise KeyError(modname)
exceptions.KeyError: 'pyspacetests2'
===============================================================================
[ERROR]: twisted.test.test_modules.ZipPathModificationTest.test_listingModules

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 163, in test_listingModules
    self._listModules()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 152, in _listModules
    pkginfo = modules.getModule(self.packageName)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 587, in __getitem__
    raise KeyError(modname)
exceptions.KeyError: 'pyspacetests4'
===============================================================================
[ERROR]: twisted.test.test_modules.ZipPathModificationTest.test_underUnderPathAlreadyImported

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 138, in test_underUnderPathAlreadyImported
    self._underUnderPathTest()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_modules.py", line 127, in _underUnderPathTest
    modinfo = modules.getModule(self.packageName)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 587, in __getitem__
    raise KeyError(modname)
exceptions.KeyError: 'pyspacetests6'
-------------------------------------------------------------------------------
```

These tests pass with all the other reactors.

<details><summary>Searchable metadata</summary>

```
trac-id__2139 2139
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1159654507000000 1159654507000000
changetime__1166851907000000 1166851907000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

They seem to be passing on the buildbot.  What configuration did they fail in?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

All my desktop machines

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

gtk/!__init!__.py rebinds sys.path to a new list object!!

I reported this upstream and a fix has been committed, so I don't think we need to worry about this anymore.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

They're failing on all of _my_ desktop machines now.  I hate to add something to the 2.5 milestone myself, but it looks like we're not going to be able to get our buildbots going again without this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Ready for review in ```rebind-2139```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @glyph|
|-|-|

Fix looks good, tests pass on my desktop and the win32 2.5 slave.  Two complaints, though:

  * Parameters to PythonPath.!__init!__ are undocumented.  Some of them get kinda-sorta documented by the class's docstring.  sysPathFactory doesn't, though.
  * the existing docstring formatting in this area is _gross_.  As long as you're adding docs for sysPathFactory to PythonPath.!__init!__'s docstring, mind reformatting it? :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [19017]) Allow twisted.python.modules to interoperate with rebinding of sys.path.

The version of pygtk currently in dapper happens to re-bind sys.path as part of
being imported, which was making it impossible to run the tests for
twisted.python.modules with the gtk2 reactor on any current dapper machine.
The gtk devs are aware of this and are apparently fixing it in a future
version, but it was blocking the task of setting up a new buildbot on a recent
OS.

More importantly, it really _is_ permissable, according to the semantics of
Python's import mechanism anyway, to rebind sys.path at any time.
twisted.python.modules' job is, in fact, to reflect those semantics accurately
so this is a necessary change.

This change also corrects some bizarre formatting problems and documentation
deficiencies in twisted.python.modules.

Author: glyph

Reviewer: exarkun

Fixes [#2139](https://github.com/twisted/twisted/issues/2139)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
